### PR TITLE
Use ASCII symbols for control commands

### PR DESCRIPTION
**Pull request by ShishkinDmitriy**
As continue of Discussion #195

### CODE REVIEW

File: src/utils/constants.h [hunk 1]
```diff
@@ -19,6 +19,17 @@ const byte MENU_ITEM_END_OF_MENU = 8;
 const byte MENU_ITEM_LIST = 9;
 const byte MENU_ITEM_PROGRESS = 10;
 //
+// Control codes
+//
+#define BACKSPACE 8   // Backspace
+#define ENTER     10  // Enter
+#define BACK      27  // Escape
```
Comment (forntoh) [inline src/utils/constants.h:26]: Does this behave the same way as on a mac?
I noticed that the mac keyboard has different encodings.
I will test this. Just commenting for your knowledge.

Comment (ShishkinDmitriy) [reply src/utils/constants.h:26]: Should be the same.
The difference is that mac and linux send `\n` (10), and windows sends `\r\n` (13 10).

File: src/input/KeyboardAdapter.h [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    /**
+     * Reset to initial state.
+     */
+    void reset();
+
+  public:
+    KeyboardAdapter(LcdMenu* menu, Stream* stream)
+        : InputInterface(menu), stream(stream) {
+    }
+    void loop() override;
```
Comment (forntoh) [inline src/input/KeyboardAdapter.h:109]:
```suggestion
    void observe() override;
```

File: src/utils/constants.h (hunk 1 above)
Comment (forntoh) [reply src/utils/constants.h]: I just checked now and pressing the `ENTER` key prints out. So it's not the same.
`Escape` is detected properly.

```
#CMD# INPUT=13
#CMD# Call=13
```

File: src/utils/constants.h [hunk 3]
```diff
@@ -19,6 +19,17 @@ const byte MENU_ITEM_END_OF_MENU = 8;
 const byte MENU_ITEM_LIST = 9;
 const byte MENU_ITEM_PROGRESS = 10;
 //
+// Control codes
+//
+#define BACKSPACE 8  // Backspace
```
Comment (forntoh) [inline src/utils/constants.h:24]: MacBook prints

```
#CMD# INPUT=127
#CMD# Call=127
```

for Backspace

Comment (ShishkinDmitriy) [reply src/utils/constants.h:24]: 127 is DEL in ASCII, so it's also valid. Both 8 and 127 will be mapped to `BACKSPACE`. Will add it as well.

File: src/utils/constants.h (hunk 1 above)
Comment (ShishkinDmitriy) [reply src/utils/constants.h]: It was unexpected) 
https://stackoverflow.com/questions/38688516/line-break-represented-as-r-on-mac
Looks like need to handle all 3 combinations.

Comment (forntoh) [reply src/utils/constants.h]: Yea, but it's actually okay, cuz it's a completely different key, the key is labeled `return` which matches the code 13 (Carriage Return)

[Apple always going their own direction 😄]

File: docs/before.puml
Comment (forntoh) [inline docs/before.puml:1]: Keep only the new `puml` (also maybe give it a new name [classes.puml](https://github.com/forntoh/LcdMenu/pull/188/files#diff-829a3e9feba91b93c220e6bbc52331e9be065ebabf45871fb6deadbeb3b58688) 🤷‍♂️) the old one will become irrelevant.

**Review by forntoh — CHANGES_REQUESTED** [review]

File: docs/classes.puml [hunk 4]
```diff
@@ -0,0 +1,49 @@
+@startuml
+' https://www.plantuml.com/plantuml/uml/
+
+hide empty members
+
+class LcdMenu {
+    +virtual bool handle(char c)
```
Comment (forntoh) [inline docs/classes.puml:7]:
```suggestion
    +virtual bool process(char c)
```

File: src/input/KeyboardAdapter.h [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+     * @brief Reset to initial state.
+     */
+    inline void reset() {
+        codeSet = CodeSet::C0;
+        csiBufferCursor = 0;
+        lastChar = 0;
+        lastCharTimestamp = 0;
+    }
+    inline bool hasLastChar() {
+        // TODO: Check millis() overflow situation
```
Comment (forntoh) [inline src/input/KeyboardAdapter.h:122]: In follow up PR or here?

Comment (ShishkinDmitriy) [reply src/input/KeyboardAdapter.h:122]: Not sure, just marked as possible problem place for now.

**Review by forntoh — APPROVED** [review]

**Pull request merged**